pipeline_dp/dp_engine.py

```diff
@@ -49,3 +49,45 @@ def aggregate(self, col, params: AggregateParams,
     # TODO: implement aggregate().
     # It returns input for now, just to ensure that the an example works.
     return col
+
+  def _bound_cross_partition_contributions(self, col,
+                                           max_partitions_contributed: int,
+                                           max_contributions_per_partition: int,
+                                           aggregator_fn):
+    """
+    Bounds the contribution by privacy_id in and cross partitions
+    Args:
+      col: collection, with types of each element: (privacy_id,
+        partition_key, value)
+      max_partitions_contributed: maximum number of partitions that one
+        privacy id can contribute to
+      max_contributions_per_partition: maximum number of records that one
+        privacy id can contribute to one partition
+      aggregator_fn: function that takes a list of values and returns an
+        aggregator object which handles all aggregation logic.
+
+    return: collection with elements ((privacy_id, partition_key),
+          aggregator)
+    """
+    # per partition-contribution bounding with bounding of each contribution
+    col = self._ops.map_tuple(col, lambda pid, pk, v: ((pid, pk), v),
+                              "To (privacy_id, partition_key), value))")
+    col = self._ops.sample_fixed_per_key(col,
+                                         max_contributions_per_partition,
+                                         "Sample per (privacy_id, partition_key)")
+    # ((privacy_id, partition_key), [value])
+    col = self._ops.map(col, lambda pid_pk: (pid_pk[0], aggregator_fn(
```

Comment:
> Please apply aggregator_fn per-partition contribution bounding, no need to do any aggregation after cross-partition bounding
> 
> Clarification:
> eventually we need to aggregate per pk, we do it in 2 steps:
> 1.Aggregate per (pid, pk) (after per-partition contributions, this PR)
> 2.Aggregate per pk (after cross-partition contributions), but that's outside of the scope of this task
> 
> Aggregation in 2 is slightly more complicated, because in 1 we can assume that data per key is in memory (because we've done bounding), but in 2 we can't that assume, so it's needed to use group by key.

That makes sense! Thanks!
Moved it to after per partition

Reply:
Thanks!